include/deal.II/rol/vector_adaptor.h

```diff
@@ -0,0 +1,478 @@
+//-----------------------------------------------------------
+//
+//    Copyright (C) 2017 by the deal.II authors
+//
+//    This file is part of the deal.II library.
+//
+//    The deal.II library is free software; you can use it, redistribute
+//    it, and/or modify it under the terms of the GNU Lesser General
+//    Public License as published by the Free Software Foundation; either
+//    version 2.1 of the License, or (at your option) any later version.
+//    The full text of the license can be found in the file LICENSE at
+//    the top level of the deal.II distribution.
+//
+//---------------------------------------------------------------
+
+#ifndef dealii_rol_vector_adaptor_h
+#define dealii_rol_vector_adaptor_h
+
+#include <deal.II/base/config.h>
+
+#ifdef DEAL_II_WITH_TRILINOS
+#include <ROL_Vector.hpp>
+
+#include <deal.II/base/exceptions.h>
+#include <deal.II/lac/vector.h>
+
+#include <type_traits>
+
+
+DEAL_II_NAMESPACE_OPEN
+
+
+using namespace dealii;
+
+
+/**
+ * A namespace that provides an interface to the
+ * <a href="https://trilinos.org/docs/dev/packages/rol/doc/html/index.html">
+ * Rapid Optimization Library</a> (ROL).
+ */
+namespace rol
+{
+
+  /**
+   * Vector adaptor that provides <tt>VectorType</tt> implementation of the
+   * ROL::Vector interface.
+   *
+   * VectorAdaptor supports vectors that satisfies the following requirements.
+   *
+   * The VectorType should contain the following types.
+   * @code
+   * VectorType::size_type;  // The type for size of the vector.
+   * VectorType::value_type; // The type for elements stored in the vector.
+   * VectorType::real_type;  // The type for real-valued numbers.
+   * @endcode
+   *
+   * However, ROL doesn't distinguish VectorAdaptor::value_type from
+   * VectorAdaptor::real_type. This is due to ROL's assumption that the
+   * VectorAdaptor::value_type itself is a type for real-valued numbers.
+   * Therefore, VectorAdaptor supports vectors whose real_type is
+   * convertible to value_type in the sense that
+   * `std::is_convertible<real_type, value_type>::value` yields `true`.
+   *
+   * The VectorType should contain the following methods.
+   * @code
+   *                   // Reinitialize the current vector using a given vector's
```

Comment:
please adjust indentation here, i guess you can remove all white spaces here and below.